---
title: "Structural Bioinformatics"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##PDB Statistics
Download a CSV file from the PDB site (accessible from “Analyze” > “PDB Statistics” > “by Experimental Method and Molecular Type”. Move this CSV file into your RStudio project and use it to answer the following questions:

> Q: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

```{r}
stats <- read.csv("Data Export Summary.csv", row.names = 1)
stats
```
```{r}
ans <- round(stats$Total / sum(stats$Total) *100, 2)
names(ans) <- rownames(stats)
ans
```


> Q: What proportion of structures in the PDB are protein?

```{r}
prot <- round(sum(stats$Proteins)/sum(stats$Total) * 100, 2)
prot
```


> Q: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

## 2. Using bio3d to analyze protein structure data.
Use the bio3d package from CRAN.

```{r}
library(bio3d)

pdb <- read.pdb("1hsg")
pdb
```

Q: How many amino acid residues are there in this pdb object? 
198

Q: Name one of the two non-protein residues? 
MK1

Q: How many protein chains are in this structure? 

## 3.

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
aa
```

# Blast or hmmer search 
# Plot a summary of search results
# List out some 'top hits'

```{r}
b <- blast.pdb(aa)
hits <- plot(b)
head(hits$pdb.id)
```
```{r}
hits <- NULL
hits$pdb.id <- c('1AKE_A','4X8M_A','6S36_A','6RZE_A','4X8H_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','4NP6_A','3GMT_A','4PZL_A')
```

```{r}
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```

```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
plot(pdbs, labels=ids)
```

```{r}
## We can view our superposed results with the new bio3d.view view() function:
library(bio3d.view)
library(rgl)

view(pdbs)
rglwidget(elementId = "tmp")
```

```{r}
anno <- pdb.annotate(ids)
unique(anno$source)
```

```{r}
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```

```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
view(pc1, col=vec2color( rmsf(pc1) ))
rglwidget(elementId = "tmp")
```

```{r}
#Plotting results with ggplot2
library(ggplot2)
library(ggrepel)

df <- data.frame(x=pc.xray$z[,1], y=pc.xray$z[,2])
col <- as.factor(grps.rd)

p <- ggplot(df, aes(x, y)) +
 geom_point(aes(col=col), size=2) +
 xlab("PC1") +
 ylab("PC2") +
 scale_color_discrete(name="Clusters") +
 geom_text_repel(aes(label=ids))
p

```


